{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"11_A_J_Geocoder.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"code","metadata":{"id":"hs-z5d7qnamp"},"source":["import numpy as np\n","import pandas as pd"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aieCF83yo07d","outputId":"970c6074-623b-461f-ef8c-da38b0ed8fc0"},"source":["from google.colab import drive\n","drive.mount('drive')"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Mounted at drive\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":204},"id":"uA4Z0J-BnepW","outputId":"0819100a-efb1-4220-fe86-51e403572880"},"source":["df = pd.read_csv('a_outliers_locations.csv')\n","df.head()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>location</th>\n","      <th>actual_category</th>\n","      <th>hours</th>\n","      <th>date_opened</th>\n","      <th>time_opened</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Ontario, Canada</td>\n","      <td>Desktop or laptop</td>\n","      <td>0.083333</td>\n","      <td>12/18/2019</td>\n","      <td>15:02</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Vienna, Italy</td>\n","      <td>Desktop or laptop</td>\n","      <td>192.383333</td>\n","      <td>10/2/2020</td>\n","      <td>11:56</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Infrastructure services</td>\n","      <td>408.183333</td>\n","      <td>10/8/2019</td>\n","      <td>13:36</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Printers and peripherals</td>\n","      <td>24.950000</td>\n","      <td>12/16/2019</td>\n","      <td>14:27</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Application support</td>\n","      <td>0.533333</td>\n","      <td>1/9/2020</td>\n","      <td>16:55</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["          location           actual_category  ...  date_opened time_opened\n","0  Ontario, Canada         Desktop or laptop  ...   12/18/2019       15:02\n","1    Vienna, Italy         Desktop or laptop  ...    10/2/2020       11:56\n","2    Sunnyvale, CA   Infrastructure services  ...    10/8/2019       13:36\n","3    Sunnyvale, CA  Printers and peripherals  ...   12/16/2019       14:27\n","4    Sunnyvale, CA       Application support  ...     1/9/2020       16:55\n","\n","[5 rows x 5 columns]"]},"metadata":{"tags":[]},"execution_count":2}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":204},"id":"ypysCRi8ntwU","outputId":"22c9aef0-7c75-44f3-e8a5-1763acd39954"},"source":["df2 = pd.read_csv('j_outliers_locations.csv')\n","df2.head()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>location</th>\n","      <th>actual_category</th>\n","      <th>hours</th>\n","      <th>date_opened</th>\n","      <th>time_opened</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Desktop or laptop</td>\n","      <td>349.682500</td>\n","      <td>12/23/2019</td>\n","      <td>13:21</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Westford, MA</td>\n","      <td>Desktop or laptop</td>\n","      <td>785.021111</td>\n","      <td>11/20/2019</td>\n","      <td>10:00</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Desktop or laptop</td>\n","      <td>276.504722</td>\n","      <td>6/17/2020</td>\n","      <td>14:32</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Westford, MA</td>\n","      <td>Desktop or laptop</td>\n","      <td>253.654167</td>\n","      <td>4/14/2020</td>\n","      <td>13:25</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Bridgewater, MA</td>\n","      <td>Desktop or laptop</td>\n","      <td>6.114444</td>\n","      <td>12/20/2019</td>\n","      <td>10:22</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["          location    actual_category       hours date_opened time_opened\n","0    Sunnyvale, CA  Desktop or laptop  349.682500  12/23/2019       13:21\n","1     Westford, MA  Desktop or laptop  785.021111  11/20/2019       10:00\n","2    Sunnyvale, CA  Desktop or laptop  276.504722   6/17/2020       14:32\n","3     Westford, MA  Desktop or laptop  253.654167   4/14/2020       13:25\n","4  Bridgewater, MA  Desktop or laptop    6.114444  12/20/2019       10:22"]},"metadata":{"tags":[]},"execution_count":3}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":409},"id":"_SEt0eyfnuEV","outputId":"10ce7570-2cf0-45c0-d8f6-a3ec6fb7582c"},"source":["#From tutoral on https://towardsdatascience.com/\n","from geopy.geocoders import Nominatim\n","from geopy.extra.rate_limiter import RateLimiter\n","locator = Nominatim(user_agent='myGeocoder')\n","geocode = RateLimiter(locator.geocode, min_delay_seconds=1)\n","df['location2'] = df['location'].apply(geocode)\n","df['point'] = df['location2'].apply(lambda loc: tuple(loc.point) if loc else None)\n","df[['latitude', 'longitude', 'altitude']] = pd.DataFrame(df['point'].tolist(), index=df.index)\n","df.to_csv('a_outlier_data.csv')\n","!cp a_outlier_data.csv \"drive/My Drive/\"\n"],"execution_count":null,"outputs":[{"output_type":"error","ename":"KeyboardInterrupt","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-7-f1504316df41>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mgeocode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRateLimiter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlocator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgeocode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin_delay_seconds\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# 2- - create location column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'location2'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'location'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgeocode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;31m# 3 - create longitude, laatitude and altitude from location column (returns tuple)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'point'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'location2'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpoint\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mloc\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.6/dist-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   4210\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4211\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4212\u001b[0;31m                 \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4213\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4214\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    128\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_last_try\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_is_last_gen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_retries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    129\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 130\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    131\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mGeocoderServiceError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_last_try\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\u001b[0m in \u001b[0;36mgeocode\u001b[0;34m(self, query, exactly_one, timeout, limit, addressdetails, language, geometry, extratags)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    306\u001b[0m         return self._parse_json(\n\u001b[0;32m--> 307\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_geocoder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexactly_one\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    308\u001b[0m         )\n\u001b[1;32m    309\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\u001b[0m in \u001b[0;36m_call_geocoder\u001b[0;34m(self, url, timeout, raw, requester, deserializer, **kwargs)\u001b[0m\n\u001b[1;32m    342\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    343\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 344\u001b[0;31m             \u001b[0mpage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrequester\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    345\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    346\u001b[0m             message = (\n","\u001b[0;32m/usr/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[1;32m    524\u001b[0m             \u001b[0mreq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmeth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    525\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 526\u001b[0;31m         \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    527\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    528\u001b[0m         \u001b[0;31m# post-process response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36m_open\u001b[0;34m(self, req, data)\u001b[0m\n\u001b[1;32m    542\u001b[0m         \u001b[0mprotocol\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    543\u001b[0m         result = self._call_chain(self.handle_open, protocol, protocol +\n\u001b[0;32m--> 544\u001b[0;31m                                   '_open', req)\n\u001b[0m\u001b[1;32m    545\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    546\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36m_call_chain\u001b[0;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[1;32m    502\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhandler\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhandlers\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    503\u001b[0m             \u001b[0mfunc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandler\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmeth_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 504\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    505\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    506\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mhttps_open\u001b[0;34m(self, req)\u001b[0m\n\u001b[1;32m   1366\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mhttps_open\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1367\u001b[0m             return self.do_open(http.client.HTTPSConnection, req,\n\u001b[0;32m-> 1368\u001b[0;31m                 context=self._context, check_hostname=self._check_hostname)\n\u001b[0m\u001b[1;32m   1369\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1370\u001b[0m         \u001b[0mhttps_request\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAbstractHTTPHandler\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdo_request_\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/urllib/request.py\u001b[0m in \u001b[0;36mdo_open\u001b[0;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[1;32m   1323\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1324\u001b[0m                 h.request(req.get_method(), req.selector, req.data, headers,\n\u001b[0;32m-> 1325\u001b[0;31m                           encode_chunked=req.has_header('Transfer-encoding'))\n\u001b[0m\u001b[1;32m   1326\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# timeout error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1327\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mURLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1279\u001b[0m                 encode_chunked=False):\n\u001b[1;32m   1280\u001b[0m         \u001b[0;34m\"\"\"Send a complete request to the server.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1281\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_send_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencode_chunked\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1282\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1283\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_send_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencode_chunked\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36m_send_request\u001b[0;34m(self, method, url, body, headers, encode_chunked)\u001b[0m\n\u001b[1;32m   1325\u001b[0m             \u001b[0;31m# default charset of iso-8859-1.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1326\u001b[0m             \u001b[0mbody\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_encode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'body'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1327\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendheaders\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbody\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencode_chunked\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencode_chunked\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1328\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1329\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mendheaders\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1274\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1275\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mCannotSendHeader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1276\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_send_output\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage_body\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mencode_chunked\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mencode_chunked\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1277\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1278\u001b[0m     def request(self, method, url, body=None, headers={}, *,\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36m_send_output\u001b[0;34m(self, message_body, encode_chunked)\u001b[0m\n\u001b[1;32m   1040\u001b[0m         \u001b[0mmsg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mb\"\\r\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_buffer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1041\u001b[0m         \u001b[0;32mdel\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_buffer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1042\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1043\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1044\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmessage_body\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m    978\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msock\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    979\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mauto_open\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 980\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    981\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    982\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mNotConnected\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mconnect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1432\u001b[0m             \u001b[0;34m\"Connect to a host on a given (SSL) port.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1433\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1434\u001b[0;31m             \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1435\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1436\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_tunnel_host\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/http/client.py\u001b[0m in \u001b[0;36mconnect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    950\u001b[0m         \u001b[0;34m\"\"\"Connect to the host and port specified in __init__.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    951\u001b[0m         self.sock = self._create_connection(\n\u001b[0;32m--> 952\u001b[0;31m             (self.host,self.port), self.timeout, self.source_address)\n\u001b[0m\u001b[1;32m    953\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msetsockopt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mIPPROTO_TCP\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msocket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTCP_NODELAY\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    954\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/socket.py\u001b[0m in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address)\u001b[0m\n\u001b[1;32m    702\u001b[0m     \u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maddress\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    703\u001b[0m     \u001b[0merr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 704\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgetaddrinfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSOCK_STREAM\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    705\u001b[0m         \u001b[0maf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msocktype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcanonname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    706\u001b[0m         \u001b[0msock\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/lib/python3.6/socket.py\u001b[0m in \u001b[0;36mgetaddrinfo\u001b[0;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[1;32m    743\u001b[0m     \u001b[0;31m# and socket type values to enum constants.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    744\u001b[0m     \u001b[0maddrlist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 745\u001b[0;31m     \u001b[0;32mfor\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32min\u001b[0m \u001b[0m_socket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetaddrinfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mport\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfamily\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    746\u001b[0m         \u001b[0maf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msocktype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcanonname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    747\u001b[0m         addrlist.append((_intenum_converter(af, AddressFamily),\n","\u001b[0;31mKeyboardInterrupt\u001b[0m: "]}]},{"cell_type":"code","metadata":{"id":"3BWRUYlE9UiU"},"source":["locator = Nominatim(user_agent='myGeocoder')\n","geocode = RateLimiter(locator.geocode, min_delay_seconds=1)\n","df2['location2'] = df2['location'].apply(geocode)\n","df2['point'] = df2['location2'].apply(lambda loc: tuple(loc.point) if loc else None)\n","df2[['latitude', 'longitude', 'altitude']] = pd.DataFrame(df2['point'].tolist(), index=df2.index)\n","df2.to_csv('j_outlier_data.csv')\n","!cp j_outlier_data.csv \"drive/My Drive/\""],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":204},"id":"dDbjiMeG_aJE","outputId":"ce1ed939-f2df-424a-f9b3-5d0c5fa84107"},"source":["df3 = pd.read_csv('comp_j_locations.csv')\n","df3.head()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>location</th>\n","      <th>actual_category</th>\n","      <th>hours</th>\n","      <th>date_opened</th>\n","      <th>time_opened</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Desktop or laptop</td>\n","      <td>44.666667</td>\n","      <td>6/9/2020</td>\n","      <td>20:46</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>Westford, MA</td>\n","      <td>Easy connect</td>\n","      <td>6.533333</td>\n","      <td>2/4/2020</td>\n","      <td>7:46</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>Houston, TX</td>\n","      <td>Desktop or laptop</td>\n","      <td>255.933333</td>\n","      <td>5/27/2020</td>\n","      <td>11:06</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Taikoo Shing, Hong Kong</td>\n","      <td>Desktop or laptop</td>\n","      <td>527.883333</td>\n","      <td>9/7/2020</td>\n","      <td>3:46</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Sunnyvale, CA</td>\n","      <td>Application support</td>\n","      <td>0.100000</td>\n","      <td>9/2/2020</td>\n","      <td>3:38</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                  location      actual_category  ...  date_opened time_opened\n","0            Sunnyvale, CA    Desktop or laptop  ...     6/9/2020       20:46\n","1             Westford, MA         Easy connect  ...     2/4/2020        7:46\n","2              Houston, TX    Desktop or laptop  ...    5/27/2020       11:06\n","3  Taikoo Shing, Hong Kong    Desktop or laptop  ...     9/7/2020        3:46\n","4            Sunnyvale, CA  Application support  ...     9/2/2020        3:38\n","\n","[5 rows x 5 columns]"]},"metadata":{"tags":[]},"execution_count":10}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"dcuBWztU_rCV","outputId":"085226e2-ae00-4579-8b46-410255efbefe"},"source":["locator = Nominatim(user_agent='myGeocoder')\n","geocode = RateLimiter(locator.geocode, min_delay_seconds=1)\n","df3['location2'] = df3['location'].apply(geocode)\n","df3['point'] = df3['location2'].apply(lambda loc: tuple(loc.point) if loc else None)\n","df3[['latitude', 'longitude', 'altitude']] = pd.DataFrame(df3['point'].tolist(), index=df3.index)\n","df3.to_csv('j1_data.csv')\n","!cp j1_data.csv \"drive/My Drive/\""],"execution_count":null,"outputs":[{"output_type":"stream","text":["RateLimiter caught an error, retrying (0/2 tries). Called with (*('Sunnyvale, CA',), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1325, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.6/http/client.py\", line 1281, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1327, in _send_request\n","    self.endheaders(body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1276, in endheaders\n","    self._send_output(message_body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1042, in _send_output\n","    self.send(msg)\n","  File \"/usr/lib/python3.6/http/client.py\", line 980, in send\n","    self.connect()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1442, in connect\n","    server_hostname=server_hostname)\n","  File \"/usr/lib/python3.6/ssl.py\", line 407, in wrap_socket\n","    _context=self, _session=session)\n","  File \"/usr/lib/python3.6/ssl.py\", line 817, in __init__\n","    self.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 1077, in do_handshake\n","    self._sslobj.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 689, in do_handshake\n","    self._sslobj.do_handshake()\n","socket.timeout: _ssl.c:835: The handshake operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1327, in do_open\n","    raise URLError(err)\n","urllib.error.URLError: <urlopen error _ssl.c:835: The handshake operation timed out>\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 367, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*('Sunnyvale, CA',), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1328, in do_open\n","    r = h.getresponse()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n","    response.begin()\n","  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n","    version, status, reason = self._read_status()\n","  File \"/usr/lib/python3.6/http/client.py\", line 272, in _read_status\n","    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n","  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n","    return self._sock.recv_into(b)\n","  File \"/usr/lib/python3.6/ssl.py\", line 1012, in recv_into\n","    return self.read(nbytes, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 874, in read\n","    return self._sslobj.read(len, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 631, in read\n","    v = self._sslobj.read(len, buffer)\n","socket.timeout: The read operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 371, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*('Sunnyvale, VA',), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1328, in do_open\n","    r = h.getresponse()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n","    response.begin()\n","  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n","    version, status, reason = self._read_status()\n","  File \"/usr/lib/python3.6/http/client.py\", line 272, in _read_status\n","    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n","  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n","    return self._sock.recv_into(b)\n","  File \"/usr/lib/python3.6/ssl.py\", line 1012, in recv_into\n","    return self.read(nbytes, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 874, in read\n","    return self._sslobj.read(len, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 631, in read\n","    v = self._sslobj.read(len, buffer)\n","socket.timeout: The read operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 371, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*(nan,), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1328, in do_open\n","    r = h.getresponse()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n","    response.begin()\n","  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n","    version, status, reason = self._read_status()\n","  File \"/usr/lib/python3.6/http/client.py\", line 272, in _read_status\n","    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n","  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n","    return self._sock.recv_into(b)\n","  File \"/usr/lib/python3.6/ssl.py\", line 1012, in recv_into\n","    return self.read(nbytes, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 874, in read\n","    return self._sslobj.read(len, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 631, in read\n","    v = self._sslobj.read(len, buffer)\n","socket.timeout: The read operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 371, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*(nan,), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1325, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.6/http/client.py\", line 1281, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1327, in _send_request\n","    self.endheaders(body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1276, in endheaders\n","    self._send_output(message_body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1042, in _send_output\n","    self.send(msg)\n","  File \"/usr/lib/python3.6/http/client.py\", line 980, in send\n","    self.connect()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1442, in connect\n","    server_hostname=server_hostname)\n","  File \"/usr/lib/python3.6/ssl.py\", line 407, in wrap_socket\n","    _context=self, _session=session)\n","  File \"/usr/lib/python3.6/ssl.py\", line 817, in __init__\n","    self.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 1077, in do_handshake\n","    self._sslobj.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 689, in do_handshake\n","    self._sslobj.do_handshake()\n","socket.timeout: _ssl.c:835: The handshake operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1327, in do_open\n","    raise URLError(err)\n","urllib.error.URLError: <urlopen error _ssl.c:835: The handshake operation timed out>\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 367, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*(nan,), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1328, in do_open\n","    r = h.getresponse()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1373, in getresponse\n","    response.begin()\n","  File \"/usr/lib/python3.6/http/client.py\", line 311, in begin\n","    version, status, reason = self._read_status()\n","  File \"/usr/lib/python3.6/http/client.py\", line 272, in _read_status\n","    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n","  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n","    return self._sock.recv_into(b)\n","  File \"/usr/lib/python3.6/ssl.py\", line 1012, in recv_into\n","    return self.read(nbytes, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 874, in read\n","    return self._sslobj.read(len, buffer)\n","  File \"/usr/lib/python3.6/ssl.py\", line 631, in read\n","    v = self._sslobj.read(len, buffer)\n","socket.timeout: The read operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 371, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n","RateLimiter caught an error, retrying (0/2 tries). Called with (*(nan,), **{}).\n","Traceback (most recent call last):\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1325, in do_open\n","    encode_chunked=req.has_header('Transfer-encoding'))\n","  File \"/usr/lib/python3.6/http/client.py\", line 1281, in request\n","    self._send_request(method, url, body, headers, encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1327, in _send_request\n","    self.endheaders(body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1276, in endheaders\n","    self._send_output(message_body, encode_chunked=encode_chunked)\n","  File \"/usr/lib/python3.6/http/client.py\", line 1042, in _send_output\n","    self.send(msg)\n","  File \"/usr/lib/python3.6/http/client.py\", line 980, in send\n","    self.connect()\n","  File \"/usr/lib/python3.6/http/client.py\", line 1442, in connect\n","    server_hostname=server_hostname)\n","  File \"/usr/lib/python3.6/ssl.py\", line 407, in wrap_socket\n","    _context=self, _session=session)\n","  File \"/usr/lib/python3.6/ssl.py\", line 817, in __init__\n","    self.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 1077, in do_handshake\n","    self._sslobj.do_handshake()\n","  File \"/usr/lib/python3.6/ssl.py\", line 689, in do_handshake\n","    self._sslobj.do_handshake()\n","socket.timeout: _ssl.c:835: The handshake operation timed out\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 344, in _call_geocoder\n","    page = requester(req, timeout=timeout, **kwargs)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 526, in open\n","    response = self._open(req, data)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 544, in _open\n","    '_open', req)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 504, in _call_chain\n","    result = func(*args)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1368, in https_open\n","    context=self._context, check_hostname=self._check_hostname)\n","  File \"/usr/lib/python3.6/urllib/request.py\", line 1327, in do_open\n","    raise URLError(err)\n","urllib.error.URLError: <urlopen error _ssl.c:835: The handshake operation timed out>\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/extra/rate_limiter.py\", line 130, in __call__\n","    return self.func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py\", line 307, in geocode\n","    self._call_geocoder(url, timeout=timeout), exactly_one\n","  File \"/usr/local/lib/python3.6/dist-packages/geopy/geocoders/base.py\", line 367, in _call_geocoder\n","    raise GeocoderTimedOut('Service timed out')\n","geopy.exc.GeocoderTimedOut: Service timed out\n"],"name":"stderr"}]}]}